command/agent/agent_endpoint.go

```diff
@@ -563,6 +564,52 @@ func (s *HTTPServer) listServers(resp http.ResponseWriter, req *http.Request) (i
 	return peers, nil
 }
 
+type reloadResponse struct {
+	Message string `json:"message"`
+}
+
+func (s *HTTPServer) AgentReloadRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	if req.Method != http.MethodPut && req.Method != http.MethodPost {
+		return nil, CodedError(405, ErrInvalidMethod)
+	}
+
+	aclObj, err := s.ResolveToken(req)
+	if err != nil {
+		return nil, err
+	}
+	if !aclObj.AllowAgentWrite() {
+		return nil, structs.ErrPermissionDenied
+	}
+
+	currConf := s.agent.GetConfig().Copy()
+
+	newConf := DefaultConfig()
+
+	for _, path := range currConf.Files {
+		if path == "" {
+			continue
+		}
+		if cfgFromFile, err := LoadConfig(path); err != nil {
+			s.logger.Error("failed to load config", "config", path, "error", err, "path", "/v1/agent/reload", "method", req.Method)
+			return nil, CodedError(400, error.Error(err))
+		} else if cfgFromFile != nil {
+			newConf = newConf.Merge(cfgFromFile)
+		}
+	}
+
+	newConf.Files = append([]string(nil), currConf.Files...)
+
+	if err := s.agent.Reload(newConf); err != nil {
```

Comment:
If you take a look at the signal handler for reloading, you'll see that this step only reloads the agent configuration and won't push any changes to the server, client, or HTTP server configuration. We should at least have parity with the signal handler here, which calls `(*Command) handleReload`. We may need to figure out a way to extract that function so that it's reachable here.